### PR TITLE
fix copy children binary nodes data

### PR DESCRIPTION
**Pull request by rederrik**
Fixed a bug with copying binary data of children nodes while doing Transport:copy.
Had also add a check in getBinaryStream() to make sure requested stream really exists.

### CODE REVIEW

File: src/Jackalope/Transport/DoctrineDBAL/Client.php [hunk 1]
```diff
@@ -2187,6 +2187,7 @@ public function updateNode(Node $node, $srcWorkspace)
 
     /**
      * {@inheritDoc}
+     * @throws RepositoryException then no binary data found
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: s/then/when

File: src/Jackalope/Transport/DoctrineDBAL/Client.php [hunk 2]
```diff
@@ -2201,6 +2202,11 @@ public function getBinaryStream($path)
             array($nodeId, $propertyName, $this->workspaceName)
         );
 
+        if (count($data) === 0) {
+            throw new RepositoryException('No binary data found in stream');
+        }
+
+
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: extra blank line

File: src/Jackalope/Transport/DoctrineDBAL/Client.php [hunk 3]
```diff
@@ -2201,6 +2202,11 @@ public function getBinaryStream($path)
             array($nodeId, $propertyName, $this->workspaceName)
         );
 
+        if (count($data) === 0) {
+            throw new RepositoryException('No binary data found in stream');
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: what exactly does that mean? and what if i created a binary multivalue property and set it to an empty array? it seems that this is allowed...

or is this meant as sanity check for database inconsistency?

Comment (rederrik) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: Mostly sanity check.
If you run the test w/o this commit it will fail with an error
>fstat() expects parameter 1 to be resource, boolean given

because the node itself has been copied, but its binary data wasn't and BinaryStreamWrapper will get this error on init.

Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: i think the bug is further down, on line 2228 where we assume that there is exactly one stream when there are not > 1.

or is this only a problem when people copied nodes before your bugfix and now have an incomplete/inconsistent database?
i think the sanity check has to happen in Property::getBinary. that method has the `length` field of the property which in case of multivalue is an array and can thus see when it did not get the number of binary from storage that was to be expected.

does that make sense?

**Comment by dbu** [comment]
the tests have been merged now. we should sort out the error handling. maybe we need a specific test case with a multivalue binary property with an empty array to check for that edge case?

File: src/Jackalope/Transport/DoctrineDBAL/Client.php (hunk 3 above)
Comment (rederrik) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: I'm not sure, but if it's multivalue shouldn't we get $data = [[],[]] instead of [] when theres simply no data at all?

**Comment by lsmith77** [comment]
can you temporarily change the composer.json file to use dev-master for phpcr-api-tests ?

File: src/Jackalope/Transport/DoctrineDBAL/Client.php (hunk 3 above)
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: i am not exactly sure either. may i ask you to add another test with a fixture that say multivalue="true" and then has not a single <value> element? that test could go into the Read chapter.

File: composer.json [hunk 4]
```diff
@@ -27,7 +27,7 @@
     },
     "require-dev": {
         "psr/log": "~1.0",
-        "phpcr/phpcr-api-tests": "2.1.6",
+        "phpcr/phpcr-api-tests": "dev-master",
```
Comment (dbu) [inline composer.json]: NOTE: once we tagged phpcr-api-tests we need to change this to 2.1.7

File: src/Jackalope/Transport/DoctrineDBAL/Client.php (hunk 3 above)
Comment (rederrik) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: Can you show me how to correctly define multivalue property in xml fixture? I've simply added another <sv:value/> and it seems to work fine..

Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Client.php]: it should be possible like this: https://github.com/phpcr/phpcr-api-tests/blob/master/fixtures/general/base.xml#L73-L77 . in the case of empty binary, it would look like `<sv:property sv:name="multivalue" sv:type="Binary" sv:multiple="true"/>` (and then hope that the fixtures loading does not choke on that :-) )

**Comment by rederrik** [comment]
Created https://github.com/phpcr/phpcr-api-tests/pull/168 to test some multivalue operations.

**Comment by lsmith77** [comment]
see https://github.com/jackalope/jackalope-doctrine-dbal/pull/299

**Comment by lsmith77** [comment]
superseded by https://github.com/jackalope/jackalope-doctrine-dbal/pull/299
